lib/Dancer2/Core/DSL.pm

```diff
@@ -241,11 +242,26 @@ sub _normalize_route {
     my $methods = shift;
     my %args;
 
-    # Options are optional, deduce their presence from arg length.
-    # @_ = ( REGEXP, OPTIONS, CODE )
-    # or
-    # @_ = ( REGEXP, CODE )
-    @args{qw/regexp options code/} = @_ == 3 ? @_ : ( $_[0], {}, $_[1] );
+    # Options are optional, try to deduce their presence from arg length.
+    if ( @_ == 4 ) {
+        # @_ = ( NAME, REGEXP, OPTIONS, CODE )
+        # get 'foo', '/foo', { 'user_agent' => '...' }, sub {...}
+        @args{qw<name regexp options code>} = @_;
+    } elsif ( @_ == 2 ) {
+        # @_ = ( REGEXP, CODE )
+        # get '/foo', sub {...}
+        @args{qw<regexp options code>} = ( $_[0], {}, $_[1] );
+    } elsif ( @_ == 3 ) {
+        # @_ = ( REGEXP, OPTIONS, CODE )
+        # get '/foo', { 'user_agent' => '...', sub {...}
+        # @_ = ( NAME, REGEXP, CODE )
+        # get 'foo', '/foo',sub {...}
+        if (ref $_[1] eq 'HASH') {
+            @args{qw<regexp options code>} = @_;
+        } else {
+            @args{qw<name regexp code>} = @_;
```

Comment:
It looks like `options` isn't set on this path --- should it be?  I not, then it can also be removed from the `@_ == 2` case.

Reply:
I'm sorry. I'm not sure I understand this comment.

Reply:
Sorry for not being sufficiently clear!  For context, I'm using substantially the same code-review standards I do at $work :) .

If `@_ == 3 && !is_hashref($_[1])`, `$args{options}` winds up unset.  This is different from all the other cases in this function.  I don't know why there is a difference, but the fact that there's a difference stood out to me.  

Does `$args{options}` need to be set when `$app->add_route` is called?  I don't think it does, because it gets plumbed through to `Route`, where `options` is not marked as required ([ref.](https://github.com/PerlDancer/Dancer2/pull/1270/files#diff-e2c044e62f9fd3169102d5e7583838633cef4548832abad3c56a714ae28b8440L40-L45)).  However, I could be wrong.

- If we do need `$args{options}` in all cases, this line could be something like
  ```suggestion
              @args{qw<name regexp code options>} = (@_, +{});
  ```
- If we do not need `$args{options}` in all cases, then line 253 can be simplified to
  ```diff
  -        @args{qw<regexp options code>} = ( $_[0], {}, $_[1] );
  +        @args{qw<regexp code>} = ( $_[0], $_[1] );
  ```

Reply:
Ah! Now I understand! That's a great catch. I fixed it to remove the added setting of an empty `options` key. That attribute has a predicate to determine if it's available. And even if it's available, it will only be used to iterate and set values.

A change that we can also make is to always set it to empty hashref if it's not provided and then remove the predicate and the check for the predicate. Since it will only trigger a loop on existing keys and there would be no keys, it wouldn't be an issue if it's set by default to empty hashref.

Thanks for the catch and for explaining it till I got it. :)